cpp/src/arrow/compute/kernels/scalar_compare.cc

```diff
@@ -439,6 +472,329 @@ struct ScalarMinMax {
+          if (!array.MayHaveNulls() ||
+              bit_util::GetBit(array.buffers[0]->data(), array.offset + row)) {
+            const offset_type* offsets = array.GetValues<offset_type>(1);
+            const uint8_t* data = array.GetValues<uint8_t>(2, /*absolute_offset=*/0);
+            const int64_t length = offsets[row + 1] - offsets[row];
+            valid_cols[col] =
+                string_view(reinterpret_cast<const char*>(data + offsets[row]), length);
+            num_valid++;
+          } else {
+            valid_cols[col] = util::nullopt;
+          }
+        }
+      }
+
+      if (num_valid == 0 || (num_valid < batch.values.size() && !options.skip_nulls)) {
+        // We had some nulls
+        builder.UnsafeAppendNull();
+        continue;
+      }
+      util::optional<string_view> result = valid_cols.front();
+      for (size_t col = 1; col < batch.values.size(); ++col) {
+        util::optional<string_view> value = valid_cols[col];
+        if (!value) {
+          DCHECK(options.skip_nulls);
+          continue;
+        }
+        result = !result ? *value : Op::Call(*result, *value);
+      }
+      if (result) {
+        builder.UnsafeAppend(*result);
+      } else {
+        builder.UnsafeAppendNull();
+      }
+    }
+
+    std::shared_ptr<Array> string_array;
+    RETURN_NOT_OK(builder.Finish(&string_array));
+    *out = *string_array->data();
+    out->mutable_array()->type = batch[0].type();
+    DCHECK_EQ(batch.length, out->array()->length);
+    DCHECK_GE(final_size,
+              checked_cast<const ArrayType&>(*string_array).total_values_length());
+    return Status::OK();
+  }
+
+  // Compute the length of the output for the given position, or -1 if it would be null.
+  static int64_t CalculateRowSize(const ElementWiseAggregateOptions& options,
+                                  const ExecBatch& batch, const int64_t index) {
+    const auto num_args = batch.values.size();
+    int64_t final_size = 0;
+    for (size_t i = 0; i < num_args; i++) {
+      int64_t element_size = 0;
+      bool valid = true;
+      if (batch[i].is_scalar()) {
+        const auto& scalar = *batch[i].scalar();
+        valid = scalar.is_valid;
+        element_size = static_cast<int64_t>(UnboxScalar<Type>::Unbox(scalar).size());
+      } else {
+        const ArrayData& array = *batch[i].array();
+        valid = !array.MayHaveNulls() ||
+                bit_util::GetBit(array.buffers[0]->data(), array.offset + index);
+        const offset_type* offsets = array.GetValues<offset_type>(1);
+        element_size = offsets[index + 1] - offsets[index];
+      }
+      if (!valid) {
+        if (options.skip_nulls) {
+          continue;
+        }
+        return -1;
+      }
+      // Conservative estimation of the element size.
+      final_size = std::max(final_size, element_size);
+    }
+    return final_size;
+  }
+};
+
+template <typename Op>
+struct FixedSizeBinaryScalarMinMax {
```

Comment:
As expected, the core implementation of `FixedSizeBinaryScalarMinMax` is very similar to the `BinaryScalarMinMax`. I am curious if they could be generalized enough to only have a single implementation or at least derive from a base class.

Reply:
I thought the same, unfortunately there is enough divergence in how element sizes are computed, which buffer to use (since fixed size doesn't have offsets) and how we compute the address of an element that I didn't find it compelling. We'd need to extract a bunch of fragmented helpers and the amount of code saved would be minimal.

We could derive from a base class, but we only have static methods here, so it's unclear to me what that'd achieve besides a form of documentation.

Reply:
Ok, thanks for the clarification. It is fine as is.